backend/services/redis_service.py

```
@@ -0,0 +1,224 @@
+                "response_metadata": chat_data.get("response_metadata", {}),
+                "last_updated": datetime.utcnow().isoformat()
+            })
+            
+            async with redis_client.pipeline() as pipe:
+                await pipe.set(key, serialized_data)
+                await pipe.expire(key, settings.REDIS_TTL)
+                await pipe.execute()
+            
+            print(f"Successfully saved chat to Redis - key: {key}, messages: {len(chat_data.get('messages', []))}")
+        except Exception as e:
+            logger.error(f"Error saving chat to Redis: {str(e)}", exc_info=True)
```

Ruff (0.8.2)

Undefined name `logger`

(F821)